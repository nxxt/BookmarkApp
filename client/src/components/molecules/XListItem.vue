<template>
  <div
    draggable
    :class="$style.list_item"
    class="d-flex flex-row py-3"
    @dragstart="dragList($event, item)"
    @mouseover="hover = true"
    @mouseleave="hover = false"
  >
    <!-- header checkbox-->
    <div
      :class="$style.list_item_header"
      class="d-flex align-center justify-start"
    >
      <x-checkbox ref="checkbox" @input="onCheckboxInput" />
    </div>

    <!-- main content -->
    <div
      :class="$style.list_item_content"
      class="pr-1 d-flex flex-column align-start justify-space-around"
    >
      <x-list-item-title
        class="subtitle-1"
        :_id="item._id"
        :number="item.watchNumber"
        :text="item.title"
        :url="item.url"
        @click="onBookmarkTitleClick"
      />

      <div
        v-if="item.memo.length > 0"
        class="caption grey--text text--lighten-1"
      >
        <x-memo :text="item.memo" />
      </div>

      <div v-if="item.tag.length > 0">
        <x-hashtag :items="item.tag" @click="onHashTagClick" />
      </div>

      <div
        class="d-flex align-center flex-wrap caption grey--text text--lighten-1"
      >
        <template v-if="item.folder.length > 0">
          <x-folder-text :items="item.folder" @click="onFolderTextClick" />
          <span>・</span>
        </template>

        <template v-if="item.domain">
          <x-domain-text :text="item.domain" />
          <span>・</span>
        </template>

        <template v-if="item.createdAt">
          <x-date-text :text="item.createdAt" />
        </template>
      </div>
    </div>

    <!-- hover -->
    <div
      v-if="hover"
      :class="$style.list_item_hover"
      class="px-4 d-flex align-center justify-end darker"
    >
      <x-tooltip-btn
        :classes="['browser_open_btn']"
        icon
        link
        tag="a"
        :href="item.url"
        target="_blank"
        tooltip="open"
        @click="onBrowserOpenBtnClick(item._id)"
      >
        <v-icon>mdi-open-in-new</v-icon>
      </x-tooltip-btn>

      <x-tooltip-btn
        :classes="['bookmark_edit_btn']"
        tooltip="edit"
        icon
        @click="onBookmarkEditBtnClick(item._id)"
      >
        <v-icon>mdi-pencil</v-icon>
      </x-tooltip-btn>

      <x-tooltip-btn
        :classes="['bookmark_remove_btn']"
        tooltip="delete"
        icon
        @click="onBookmarkRemoveBtnClick(item._id)"
      >
        <v-icon>mdi-delete</v-icon>
      </x-tooltip-btn>
    </div>
  </div>
</template>

<script lang="ts">
import { PropType } from 'vue'
import { Component, Prop, Vue } from 'vue-property-decorator'
import IBookmark from '@@/types/IBookmark'

import XCheckbox from '@/components/atoms/XCheckbox.vue'
import XMemo from '@/components/atoms/XMemo.vue'
import XHashtag from '@/components/atoms/XHashtag.vue'
import XFolderText from '@/components/atoms/XFolderText.vue'
import XDomainText from '@/components/atoms/XDomainText.vue'
import XDateText from '@/components/atoms/XDateText.vue'
import XTooltipBtn from '@/components/atoms/XTooltipBtn.vue'
import XListItemTitle from '@/components/molecules/XListItemTitle.vue'

@Component({
  components: {
    XCheckbox,
    XMemo,
    XHashtag,
    XFolderText,
    XDomainText,
    XDateText,
    XTooltipBtn,
    XListItemTitle
  }
})
export default class XListItem extends Vue {
  hover: boolean = false

  @Prop({ type: Object as PropType<IBookmark>, default: () => ({}) })
  item!: IBookmark

  @Prop({ type: String, default: '' })
  color!: string

  // darag & drop
  dragList(event: Event & { dataTransfer?: any }, item: IBookmark) {
    event.dataTransfer.setData('bookmarkId', item._id)
  }

  // checkbox
  onCheckboxInput(isChecked: boolean): void {
    if (isChecked) this.$emit('checkbox-on', this.item._id)
    else this.$emit('checkbox-off', this.item._id)
  }

  // list content
  onBookmarkTitleClick(_id: string) {
    this.$emit('bookmark-title-click', _id)
  }

  onHashTagClick(_id: string) {
    this.$emit('hash-tag-click', _id)
  }

  onFolderTextClick(_id: string) {
    this.$emit('folder-text-click', _id)
  }

  // hover
  onBrowserOpenBtnClick(_id: string) {
    this.$emit('browser-open-btn-click', _id)
  }

  onBookmarkEditBtnClick(_id: string) {
    this.$emit('bookmark-edit-btn-click', _id)
  }

  onBookmarkRemoveBtnClick(_id: string) {
    this.$emit('bookmark-remove-btn-click', _id)
  }

  setCheck(isCheck: boolean) {
    const checkbox = this.$refs.checkbox as XCheckbox
    checkbox.setCheck(isCheck)
  }
}
</script>

<style module>
.list_item {
  position: relative;
  width: 100%;
}
.list_item_header {
  width: 30px;
}
.list_item_content {
  width: calc(100% - 30px);
}
.list_item_hover {
  position: absolute;
  top: 0;
  right: 0;
  width: auto;
  opacity: 0.8;
  height: 100%;
}
</style>
