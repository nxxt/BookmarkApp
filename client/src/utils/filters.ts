import IBookmark from '~~/types/IBookmark'
import ITag from '~~/types/ITag'
import IFolder from '~~/types/IFolder'

const filters: { [key: string]: Function } = {
  bookmarkTitle: (bookmark: IBookmark) => {
    return bookmark && bookmark.title ? bookmark.title : ''
  },
  tagName: (tag: ITag) => {
    return tag && tag.name ? `#${tag.name}` : ''
  },
  folderName: (folder: IFolder) => {
    return folder && folder.name ? folder.name : ''
  },
  textOverflow: (str: string) => {
    return str.length > 140 ? `${str.substring(0, 140)}â€¦` : str
  },
  dateFormat: (str: string) => {
    return str.substring(0, 10)
  }
}

export default filters
