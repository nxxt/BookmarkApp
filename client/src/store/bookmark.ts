import { Module, VuexModule, Mutation, Action } from 'vuex-module-decorators'
import Bookmark from '@@/types/IBookmark'
// import DtoBookmark from '@@/types/DtoBookmark'
import ISortType from '@@/types/ISortType'
import ITag from '@@/types/ITag'
import { $axios } from '@/utils/api'
// import IFolder from '~~/types/IFolder'

// stateFactory: true → Vuex をモジュールモードで扱うために指定
@Module({
  stateFactory: true,
  namespaced: true,
  name: 'bookmark'
})
export default class Bookmarks extends VuexModule {
  private all: Bookmark[] = []
  private unfoldered: Bookmark[] = []
  private display: Bookmark[] = []
  private folderId: string = ''

  get allBookmarks() {
    return this.all
  }

  get unfolderedBookmarks() {
    return this.unfoldered
  }

  get displayBookmarks() {
    return this.display
  }

  get _fId() {
    return this.folderId
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private FETCH_ALL(items: Bookmark[]) {
    this.all = items
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private FETCH_UNFOLDERED(items: Bookmark[]) {
    this.unfoldered = items
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private FETCH_DISPLAY(items: Bookmark[], folderId: string) {
    this.display = items
    this.folderId = folderId
  }

  /**
   * Bookmark を state に格納する
   * @param items Bookmark オブジェクト配列
   */
  @Mutation
  private ADD_DISPLAY(items: Bookmark[]) {
    this.display = [...this.display, ...items]
  }

  /**
   * Bookmark を追加する
   * @param bookmark Bookmark オブジェクト
   */
  @Mutation
  private CREATE(data: { bookmark: Bookmark; folderId: string }) {
    // リストに Bookmark を追加
    this.all.push(data.bookmark)
    this.unfoldered.push(data.bookmark)
    // 選択したsidemenuが「all bookmarks」「unfoldered」の場合
    if (['h-0', 'h-1'].includes(data.folderId)) {
      this.display.push(data.bookmark)
    }
  }

  /**
   * Bookmark を更新する
   * @param bookmark Bookmark オブジェクト
   */
  @Mutation
  private UPDATE(bookmark: Bookmark) {
    const target = this.all.find((d) => d._id === bookmark._id)
    if (target) {
      const index = this.all.indexOf(target)
      Object.assign(this.all[index], bookmark)
    }
    const target2 = this.unfoldered.find((d) => d._id === bookmark._id)
    if (target2) {
      const index = this.unfoldered.indexOf(target2)
      Object.assign(this.unfoldered[index], bookmark)
    }
    const target3 = this.display.find((d) => d._id === bookmark._id)
    if (target3) {
      const index = this.display.indexOf(target3)
      Object.assign(this.display[index], bookmark)
    }
  }

  /**
   * Bookmark を削除する
   * @param bookmark 削除する Bookmark インスタンス
   */
  @Mutation
  private DELETE(_id: string) {
    const target = this.all.find((d: Bookmark) => d._id === _id)
    if (target) {
      const index = this.all.indexOf(target)
      this.all.splice(index, 1)
    }
    const target2 = this.unfoldered.find((d: Bookmark) => d._id === _id)
    if (target2) {
      const index = this.unfoldered.indexOf(target2)
      this.unfoldered.splice(index, 1)
    }
    const target3 = this.display.find((d: Bookmark) => d._id === _id)
    if (target3) {
      const index = this.display.indexOf(target3)
      this.display.splice(index, 1)
    }
  }

  @Action({})
  public async fetchAllBookmarks(params: { userId: string }): Promise<void> {
    const data = await $axios.$get(`/api/v1/bookmarks/all/${params.userId}`)
    this.FETCH_ALL(data)
  }

  @Action({})
  public async fetchUnfolderedBookmarks(params: {
    userId: string
  }): Promise<void> {
    const data = await $axios.$get(
      `/api/v1/bookmarks/unfoldered/${params.userId}`
    )
    this.FETCH_UNFOLDERED(data)
  }

  @Action({})
  public async fetchDisplayBookmarks(params: {
    userId: string
    pageNo: number
    perPage: number
    folderId: string
    filter: (string | ITag)[]
    sort: ISortType
  }): Promise<any> {
    const { userId, pageNo, perPage, folderId, filter, sort } = params
    const res = await $axios.post(
      `/api/v1/bookmarks/list/${userId}?page_no=${pageNo}&per_page=${perPage}&folder_id=${folderId}`,
      { sort, filter }
    )
    this.FETCH_DISPLAY(res.data, params.folderId)
    return [res.data, Number(res.headers['x-list-total-length'])]
  }

  @Action({})
  public async addDisplayBookmarks(params: {
    userId: string
    pageNo: number
    perPage: number
    folderId: string
    filter: (string | ITag)[]
    sort: ISortType
  }): Promise<void> {
    const { userId, pageNo, perPage, folderId, filter, sort } = params

    const data = await $axios.$post(
      `/api/v1/bookmarks/list/${userId}?page_no=${pageNo}&per_page=${perPage}&folder_id=${folderId}`,
      { sort, filter }
    )
    this.ADD_DISPLAY(data)
  }

  @Action({})
  async createBookmark(data: {
    url: string
    userId: string
    folderId: string
  }) {
    const item = await $axios.$post(`/api/v1/bookmarks`, {
      url: data.url,
      userId: data.userId
    })
    this.CREATE({ bookmark: item, folderId: data.folderId })
    return item
  }

  @Action({})
  async updateBookmark(bookmark: Bookmark) {
    const item = await $axios.$patch(
      `/api/v1/bookmarks/${bookmark._id}`,
      bookmark
    )
    this.UPDATE(item)
    return item
  }

  @Action({})
  async deleteBookmark(_id: string) {
    const data = await $axios.$delete(`/api/v1/bookmarks/${_id}`)
    this.DELETE(_id)
    return data
  }
}
