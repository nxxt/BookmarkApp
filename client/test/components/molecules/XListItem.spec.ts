import Vuetify from 'vuetify'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import bookmark from '@@/test/mock_data/bookmark.json'
import XListItem from '@/components/molecules/XListItem.vue'
import filters from '@/utils/filters'
import mockGetters from '@/utils/mock_getters'

const localVue = createLocalVue()
Object.keys(filters).forEach((key: string) =>
  localVue.filter(key, filters[key])
)
Object.keys(mockGetters).forEach((key: string) => {
  localVue.prototype[key] = mockGetters[key]
})

describe('XListItem.vue', () => {
  /****************
   * 準備
   ****************/
  // vuetify
  let vuetify: typeof Vuetify
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  // propsで渡す値
  const propsData = {
    item: bookmark,
    color: 'darker'
  }

  const factory = (options?: any) => {
    return shallowMount(XListItem, {
      localVue,
      vuetify,
      ...options
    })
  }

  /****************
   * TEST
   ****************/
  it('props', () => {
    const wrapper = factory({ propsData })
    // console.log(wrapper.html())
    // VueインスタンスのpropsがpropsDataと等しいか
    expect(wrapper.props()).toStrictEqual(propsData)
  })

  describe('metohds', () => {
    it('onCheckboxInput', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const eventOn = jest.fn()
      const eventOff = jest.fn()
      wrapper.vm.$on('checkbox-on', eventOn)
      wrapper.vm.$on('checkbox-off', eventOff)

      // 初期状態
      expect(eventOn).toBeDefined() // イベントがundefinedでないこと
      expect(eventOff).toBeDefined() // イベントがundefinedでないこと
      expect(eventOn).toHaveBeenCalledTimes(0) // eventが発生していないこと
      expect(eventOff).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onCheckboxInput(true)

      // イベント後
      expect(eventOn).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('checkbox-on')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと

      // emit
      vm.onCheckboxInput(false)

      // イベント後
      expect(eventOff).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const b = wrapper.emitted('checkbox-off')
      expect(b).toBeDefined()
      if (b) expect(b[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })

    it('onBookmarkTitleClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const event = jest.fn()
      wrapper.vm.$on('bookmark-title-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onBookmarkTitleClick(wrapper.vm.$props.item._id)

      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('bookmark-title-click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })

    it('onHashTagClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const event = jest.fn()
      wrapper.vm.$on('hash-tag-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onHashTagClick(wrapper.vm.$props.item._id)

      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('hash-tag-click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })

    it('onFolderTextClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const event = jest.fn()
      wrapper.vm.$on('folder-text-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onFolderTextClick(wrapper.vm.$props.item._id)

      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('folder-text-click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })

    it('onBrowserOpenBtnClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const event = jest.fn()
      wrapper.vm.$on('browser-open-btn-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onBrowserOpenBtnClick(wrapper.vm.$props.item._id)

      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('browser-open-btn-click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })

    it('onBookmarkEditBtnClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const event = jest.fn()
      wrapper.vm.$on('bookmark-edit-btn-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onBookmarkEditBtnClick(wrapper.vm.$props.item._id)

      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('bookmark-edit-btn-click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })

    it('onBookmarkRemoveBtnClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm

      const event = jest.fn()
      wrapper.vm.$on('bookmark-remove-btn-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // emit
      vm.onBookmarkRemoveBtnClick(wrapper.vm.$props.item._id)

      // イベント後
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('bookmark-remove-btn-click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(wrapper.vm.$props.item._id) // イベントのペイロードが正しいこと
    })
  })

  describe('the entire template', () => {
    it('snapshot', () => {
      const wrapper = factory({ propsData })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
