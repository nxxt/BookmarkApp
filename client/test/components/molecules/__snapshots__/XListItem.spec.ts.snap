// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XListItem.vue the entire template snapshot 1`] = `
"<div draggable=\\"true\\" class=\\"d-flex flex-row py-3 list_item\\">
  <div class=\\"d-flex align-center justify-start list_item_header\\">
    <x-checkbox-stub></x-checkbox-stub>
  </div>
  <div class=\\"pr-1 d-flex flex-column align-start justify-space-around list_item_content\\">
    <x-list-item-title-stub _id=\\"5ef3668015ceb88170b5b001\\" number=\\"5\\" text=\\"Vue.js\\" url=\\"https://jp.vuejs.org/index.html\\" class=\\"subtitle-1\\"></x-list-item-title-stub>
    <div class=\\"caption grey--text text--lighten-1\\">
      <x-memo-stub text=\\"Vue.js - The Progressive JavaScript Framework\\"></x-memo-stub>
    </div>
    <div>
      <x-hashtag-stub items=\\"5ef3539b41c7218314tag001,5ef3522c41c7218314tag002\\"></x-hashtag-stub>
    </div>
    <div class=\\"d-flex align-center flex-wrap caption grey--text text--lighten-1\\">
      <x-folder-text-stub items=\\"5ef35e1118cb777folder001,5ef6ab9cbe8bcd4folder002\\"></x-folder-text-stub> <span>・</span>
      <x-domain-text-stub text=\\"jp.vuejs.org\\"></x-domain-text-stub> <span>・</span>
      <x-date-text-stub text=\\"2020-06-24T14:43:12.335Z\\"></x-date-text-stub>
    </div>
  </div>
  <!---->
</div>"
`;
